### PR TITLE
Type annotations

### PR DESCRIPTION
**Pull request by TylerLubeck**
Adding in PEP 484 style type annotations and a [mypy](http://mypy.readthedocs.io/en/latest/) test step.

Ticket: #712 

/cc @hackebrot @audreyr @pydanny

### CODE REVIEW

**Comment by TylerLubeck** [comment]
Calling it a night for the time being. Running in to some issues with mypy, but I'll check in with their team tomorrow.

File: cookiecutter/cli.py [hunk 1]
```diff
@@ -48,7 +51,7 @@ def validate_extra_context(ctx, param, value):
 
     # Convert tuple -- e.g.: (u'program_name=foobar', u'startsecs=66')
     # to dict -- e.g.: {'program_name': 'foobar', 'startsecs': '66'}
-    return dict(s.split('=', 1) for s in value) or None
+    return dict(s.split('=', 1) for s in value) or None  # type: ignore
```
Comment (TylerLubeck) [inline cookiecutter/cli.py]: This is sub-optimal, but relies on a fix here: https://github.com/python/typeshed/issues/256

**Comment by TylerLubeck** [comment]
I should point out that a lot of these annotations break the line-length limits. This is a [known issue][mypy-issue] with the mypy annotation parser, but it looks like it'll be fixed in mypy 0.4.2. Currently, this runs on [mypy 0.4.1](https://pypi.python.org/pypi/mypy-lang/0.4.1).


[mypy-issue]: https://github.com/python/mypy/issues/1102

File: tox.ini [hunk 2]
```diff
@@ -21,3 +21,9 @@ commands =
 
 [testenv:cov-report]
 commands = py.test --cov=cookiecutter --cov-report=term --cov-report=html
+
+[testenv:mypy]
+deps =
+    mypy-lang
+# Not all imports have stubs yet, so let's ignore those for now.
+commands = mypy cookiecutter --silent-imports
```
Comment (TylerLubeck) [inline tox.ini]: The `--silent-imports` argument is to silence type errors that come form importing non-annotated code. It's also possible to put a `# type: ignore` on each of the offending import lines, but doing it this way makes it easier to throw a switch at runtime and see if anything has changed.

**Comment by audreyr** [comment]
Awesome work so far @TylerLubeck.

For long lines, I'd suggest:
* Add a comment so the linter ignores the line (which will be temporary)
* When mypy 0.4.2 comes out, send us a PR to break up only the too-long lines, as in your example

**Comment by audreyr** [comment]
@TylerLubeck if you get busy and want us to continue where you left off, let us know. Otherwise, we look forward to the last part of your PR 😄

**Comment by TylerLubeck** [comment]
Hey @audreyr, my apologies. I'll get to this tonight!

**Comment by TylerLubeck** [comment]
All set @audreyr. If this is something you want to merge, do you want me to squash or rebase? I'm used to doing both, but I know it's a hot topic.

**Comment by willingc** [comment]
@TylerLubeck Nice work!

**Comment by TylerLubeck** [comment]
Hi all, just checking in. Anything else I can do here?

**Comment by michaeljoseph** [comment]
Looks 🍰  to me, nice job @TylerLubeck .
🔔 @audreyr

**Comment by hackebrot** [comment]
I'm sorry that this hasn't been merged yet. There are a number of conflicts now 😢 

Any volunteers for getting these resolved?

**Comment by theodesp** [comment]
@hackebrot I may look at this as I'm working on tickets this week

**Comment by themanifold** [comment]
Anyone still looking at this?

**Comment by insspb** [comment]
@TylerLubeck thank you for your efforts. Unfortunately this was not merged in time, currently, we do project cleanup and cannot merge it in the current state. 
Do you have interest to adapt it to current cookiecutter version?

File: tox.ini (hunk 2 above)
Comment (ssbarnea) [reply tox.ini]: Closing as this is far too old and also has a weird implementation. Enabling mypy can be done as part of the flake8, and not with extra tox environments. 

If anyone wants to restart the challenge, I will be happy to help.

**Pull request closed without merging**